### PR TITLE
feat: katib-db-manager sidecar rewrite

### PR DESCRIPTION
**Pull request by NohaIhab**
# Summary
- define pebble layer
- update event handling
- introduce on_remove handler
- patch the port with KubernetesServicePatch lib
- deploy k8s resources with KubernetesResourceHandler
- limit my-sql relation to 1 at a time in metadata.yaml
- update metadata.yaml
- introduce health checks and update status handler to report them

### CODE REVIEW

File: charms/katib-db-manager/metadata.yaml [hunk 1]
```diff
@@ -6,8 +6,11 @@ description: |
   Katib is the project which is agnostic to machine learning (ML) frameworks. It can tune
   hyperparameters of applications written in any language of the users’ choice and natively
   supports many ML frameworks, such as TensorFlow, MXNet, PyTorch, XGBoost, and others.
-min-juju-version: "2.9.0"
-series: [kubernetes]
+website: https://charmhub.io/katib-db-manager
+source: https://github.com/canonical/katib-operators/tree/main/charms/katib-db-manager
+containers:
```
Comment (i-chvets) [inline charms/katib-db-manager/metadata.yaml:11]: Please add `issues` filed as well:
`issues: https://github.com/canonical/katib-operators`

File: charms/katib-db-manager/metadata.yaml [hunk 2]
```diff
@@ -6,8 +6,11 @@ description: |
   Katib is the project which is agnostic to machine learning (ML) frameworks. It can tune
   hyperparameters of applications written in any language of the users’ choice and natively
   supports many ML frameworks, such as TensorFlow, MXNet, PyTorch, XGBoost, and others.
-min-juju-version: "2.9.0"
-series: [kubernetes]
+website: https://charmhub.io/katib-db-manager
+source: https://github.com/canonical/katib-operators/tree/main/charms/katib-db-manager
```
Comment (i-chvets) [inline charms/katib-db-manager/metadata.yaml:10]: Please add `issues` filed as well:
`issues: https://github.com/canonical/katib-operators/issues`

File: charms/katib-db-manager/src/charm.py [hunk 3]
```diff
@@ -4,12 +4,22 @@
...
 from ops.model import ActiveStatus, MaintenanceStatus, WaitingStatus
+from ops.pebble import Layer
 
-logger = logging.getLogger(__name__)
+K8S_RESOURCE_FILES = [
+    "src/templates/auth_manifests.yaml.j2",
+]
 
 
 class CheckFailed(Exception):
```
Comment (i-chvets) [inline charms/katib-db-manager/src/charm.py:25]: Is this used after re-write?

File: charms/katib-db-manager/tests/unit/test_operator.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    }
+    harness.update_relation_data(rel_id, mysql_unit, data)
+    harness.container_pebble_ready("katib-db-manager")
+    pebble_plan = harness.get_container_pebble_plan("katib-db-manager")
+    assert pebble_plan
+    assert pebble_plan._services
+    pebble_plan_info = pebble_plan.to_dict()
+    assert pebble_plan_info["services"]["katib-db-manager"]["command"] == "./katib-db-manager"
+    test_env = pebble_plan_info["services"]["katib-db-manager"]["environment"]
+    assert 6 == len(test_env)
```
Comment (i-chvets) [inline charms/katib-db-manager/tests/unit/test_operator.py:126]: Put comment what that `6` means:
`# there should be size (6) environment variables`

File: charms/katib-db-manager/src/charm.py (hunk 3 above)
Comment (NohaIhab) [reply charms/katib-db-manager/src/charm.py:25]: yes, it's passed to the KubernetesResourceHandler

File: tests/integration/test_charms.py [hunk 5]
```diff
@@ -44,7 +44,7 @@ async def test_deploy_katib_charms(ops_test: OpsTest):
     # Deploy katib-controller, katib-db-manager, and katib-ui charms
     await ops_test.model.deploy(controller_charm, resources={"oci-image": controller_image_path})
```
Comment (i-chvets) [inline tests/integration/test_charms.py:45]: Do we need trust here as well?

Comment (NohaIhab) [reply tests/integration/test_charms.py:45]: katib-controller is still a podspec so it's not needed

File: charms/katib-db-manager/src/charm.py (hunk 3 above)
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:25]: Hmm, I could not find `CheckFailed` to be used anywhere in `charm.py`. Am I missing something?
I thought we changed all that error handling to use `ErrorWithStatus`.

Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:25]: Got it. I was talking about `CheckFailed` class. Let's remove it.

File: charms/katib-db-manager/src/charm.py [hunk 6]
```diff
@@ -4,123 +4,137 @@
...
-                            }
-                        },
-                    }
-                ],
-            }
+        # retrieve configuration and base settings
+        self.logger = logging.getLogger(__name__)
+        self._container_name = "katib-db-manager"
+        self._container = self.unit.get_container(self._container_name)
+        self._exec_command = "./katib-db-manager"
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/src/charm.py:35]: is there a reason to put exec command here and not directly in the layer?
do we use it anywhere else?

File: charms/katib-db-manager/tox.ini [hunk 7]
```diff
@@ -60,4 +61,19 @@ commands =
...
+    coverage report
+deps =
+    -r requirements-unit.txt
+description = Run unit tests
+
+[testenv:integration]
+commands = pytest -v --tb native --asyncio-mode=auto {[vars]tst_path}integration --log-cli-level=INFO -s {posargs}
+deps =
+    -r requirements-integration.txt
+description = Run integration tests
```
Comment (beliaev-maksim) [inline charms/katib-db-manager/tox.ini:79]:
```suggestion
description = Run integration tests

```

File: charms/katib-db-manager/src/charm.py (hunk 6 above)
Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:35]: We've been doing this way for all of our charms. No particular reason, just keeping all container related config together: container name, container exec command.

Comment (beliaev-maksim) [reply charms/katib-db-manager/src/charm.py:35]: general rule in coding keep your variables close to the executable code.

this comment is not blocking the PR, but please reconsider the pattern

**Review by beliaev-maksim — APPROVED** [review]

File: charms/katib-db-manager/src/charm.py [hunk 8]
```diff
@@ -130,10 +144,110 @@ def _check_mysql(self):
...
+                # conflict detected when applying K8S resources
+                # re-apply K8S resources with forced conflict resolution
+                self.unit.status = MaintenanceStatus("Force applying K8S resources")
+                self.logger.warning("Apply K8S resources with forced changes against conflicts")
+                self.k8s_resource_handler.apply(force=force_conflicts)
+            else:
+                raise GenericCharmRuntimeError("K8S resources creation failed") from error
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
+
+    def _get_check_status(self):
```
Comment (i-chvets) [inline charms/katib-db-manager/src/charm.py:181]: Do we really need a separate function that we use only once?

Comment (NohaIhab) [reply charms/katib-db-manager/src/charm.py:181]: I needed to mock it in unit tests so I can parameterize the status

Comment (i-chvets) [reply charms/katib-db-manager/src/charm.py:181]: Got it.

**Review by i-chvets — APPROVED** [review]
Approved.

**Pull request merged**